### DISCUSSION TITLE
Shutter Auto Setup with Shelly 2.5 and shelly plus 2PM

### DISCUSSION THREAD

**Discussion by stefanbode** [arendst/Tasmota#18388] [Ideas]
Hi Team,

the shutter [ESP32] will get new functionality with the power measurement system of the Shelly devices. Prerequisite will be that the shutter has configured endstops where automatically the shutter stop and we can measure the stop with a change of current. With some evolution cycles it may become smarter over time.

First challenge to solve: Where can I get the power of channel 1 and 2 inside the code. And because we need this in a 50ms intervall what is the option call an update of the power.

**Comment by stefanbode** [comment]
With some tricks to force an update the power consumption can be found and used. Promising..
```
      if (ShutterGlobal.callibration_run) {
        // update energy consumption on every loop to dectect stop of the shutter
        XnrgCall(FUNC_ENERGY_EVERY_SECOND);
        // fency calculation with direction gives index 0 and 1 of the energy meter
        // stop if endpoint is reached
        if (Energy->active_power[(Shutter[i].direction + 1) / 2] == 0.0 && Shutter[i].time > 20){
        Shutter[i].target_position = Shutter[i].real_position;
        }
      }
```

**Comment by stefanbode** [comment]
There is pull request https://github.com/arendst/Tasmota/pull/18393 for shelly PLUS 2PM (ESP32 based) only. New command "shuttersetup". Right now not fully automated but a first step. End stops must be configured before. Stop position detected by current==0 motor goes off. If there is remaining power please let me know. Maybe it must be less than 10W or something.

**Comment by stefanbode** [comment]
Shelly plus 2.5 PM should use with buttons following template:
`{"NAME":"Shelly Plus 2PM PCB v0.1.9","GPIO":[320,0,0,0,34,128,0,0,225,224,0,0,0,0,129,0,0,0,0,0,0,608,640,3458,0,0,0,0,0,9472,0,4736,0,0,0,0],"FLAG":0,"BASE":1}`
Setup Now validated to work with the module above. Shutter must be closed before start. Then "`shuttersetup`" and openduration and closeduration will be measured and used. also `shuttersethalfway `will be set to 70 as a good initial point for most shutters.

**Comment by arendst** [comment]
@stefanbode did you have any luck in detecting shutter end-positions using ``XnrgCall(FUNC_ENERGY_EVERY_SECOND)``?

I ask because I do not expect this to be a viable solution. I think a better approach is monitoring the current and if it raises above a percentage of normal operating current this should flag the end position.

The problem with the Tasmota energy monitoring sensors is that they use different intervals for updating their values based on their hardware designs. The CSE7766 device send updates around every 100ms. The CSE7761 is polled every 200ms. The BL6523 is polled every 250ms. The HLW8012 and ADE7953 are polled every 1000ms. The ADE7880 is programmed to measure during 1000ms before sending and interupt for being polled.

So just reading the ``Energy->current[channel]`` is not an option (as you already know) as it is only updated once 1 second for ADE and HLW based devices.

Your current solution for the ShellyPlus2PM (using the ADE7953) and using ``Energy->active_power[channel]`` will result in an active power value result / 20 as you now measure the Energy in 1/20 of a second. You might have more luchkin monitoring the current in this case which seems to be stable untill the moter is at end-point.

For ADE7880 I see no easy solution simple because we programmed to sensor to measure during 50Hz cycles (= 1 second) for most practical results.

Anyway, let users play with it and keep in mind what is (not)possible in energy monitoring hardware design.

**Reply by stefanbode** [reply]
Hi Theo, you're absolute right. With the force read every 50ms you only get 1/20 of the power during this calibration run. Anyhow this is at least about 9W which is enough to detect the end stop. I had to implement a delay to detect the endstop reliable only after the SECOND "0W" that came in; because in some situations you will get a "0W" during the run. But this does not happen often. And as far as I can see nearly never two times in a sequence. The repetitive accuracy is also below 0.1sec. So the duration times are quite stable if you do it several times. I assume this is quite similar how they do it in the original shelly software. For sure their detect interval is not 1 second. All in all I'm currently fine. If there will be no additional bugs coming in next week we can make the ESP32 shutter default for this device. Will drop you a note
